lib/std/dwarf.zig

```diff
@@ -1477,3 +1906,561 @@ fn getStringGeneric(opt_str: ?[]const u8, offset: u64) ![:0]const u8 {
+    ) !void {
+        const entry_size = try entrySize(self.table_enc);
+
+        var left: usize = 0;
+        var len: usize = self.fde_count;
+
+        var stream = io.fixedBufferStream(self.entries);
+        const reader = stream.reader();
+
+        while (len > 1) {
+            const mid = left + len / 2;
+
+            try stream.seekTo(mid * entry_size);
+            const pc_begin = try readEhPointer(reader, self.table_enc, @sizeOf(usize), .{
+                .pc_rel_base = @intFromPtr(&self.entries[stream.pos]),
+                .follow_indirect = true,
+                .data_rel_base = eh_frame_hdr_ptr,
+            }, builtin.cpu.arch.endian()) orelse return badDwarf();
+
+            if (pc < pc_begin) {
+                len /= 2;
+            } else {
+                left = mid;
+                if (pc == pc_begin) break;
+                len -= len / 2;
+            }
+        }
+
+        if (len == 0) return badDwarf();
+        try stream.seekTo(left * entry_size);
+
+        // Read past the pc_begin field of the entry
+        _ = try readEhPointer(reader, self.table_enc, @sizeOf(usize), .{
+            .pc_rel_base = @intFromPtr(&self.entries[stream.pos]),
+            .follow_indirect = true,
+            .data_rel_base = eh_frame_hdr_ptr,
+        }, builtin.cpu.arch.endian()) orelse return badDwarf();
+
+        const fde_ptr = math.cast(usize, try readEhPointer(reader, self.table_enc, @sizeOf(usize), .{
+            .pc_rel_base = @intFromPtr(&self.entries[stream.pos]),
+            .follow_indirect = true,
+            .data_rel_base = eh_frame_hdr_ptr,
+        }, builtin.cpu.arch.endian()) orelse return badDwarf()) orelse return badDwarf();
+
+        // Verify the length fields of the FDE header are readable
+        if (!self.isValidPtr(fde_ptr, isValidMemory, eh_frame_len) or fde_ptr < self.eh_frame_ptr) return badDwarf();
+
+        var fde_entry_header_len: usize = 4;
+        if (!self.isValidPtr(fde_ptr + 3, isValidMemory, eh_frame_len)) return badDwarf();
+        if (self.isValidPtr(fde_ptr + 11, isValidMemory, eh_frame_len)) fde_entry_header_len = 12;
+
+        // Even if eh_frame_len is not specified, all ranges accssed are checked by isValidPtr
+        const eh_frame = @as([*]const u8, @ptrFromInt(self.eh_frame_ptr))[0 .. eh_frame_len orelse math.maxInt(u32)];
+
+        const fde_offset = fde_ptr - self.eh_frame_ptr;
+        var eh_frame_stream = io.fixedBufferStream(eh_frame);
+        try eh_frame_stream.seekTo(fde_offset);
+
+        const fde_entry_header = try EntryHeader.read(&eh_frame_stream, .eh_frame, builtin.cpu.arch.endian());
+        if (!self.isValidPtr(@intFromPtr(&fde_entry_header.entry_bytes[fde_entry_header.entry_bytes.len - 1]), isValidMemory, eh_frame_len)) return badDwarf();
+        if (fde_entry_header.type != .fde) return badDwarf();
+
+        // CIEs always come before FDEs (the offset is a subtraction), so we can assume this memory is readable
```

Comment:
This code is specifically only searching `.eh_frame_hdr` which always uses `.eh_frame`, where this is true.

For `.eh_frame` sections, the CIE offset is defined as:

> A 4 byte unsigned value that when subtracted from the offset of the current FDE yields the offset of the start of the associated CIE. This value shall never be 0.

For `.debug_frame` sections, the CIE offset of an FDE is encoded as an offset into the `.debug_frame` section, so in that case it can indeed be anywhere, including after the FDE.  `dwarf.zig:2168` is where I make this distinction in the parser.

Reply:
I think I could still trick the unwinder into looking ahead by specifying a negative offset in the CIE offset placeholder. Anyway, I don't think there exists a linker that does that including `zld` so we should be good here. Thanks for quoting the spec btw!

Reply:
Actually this is a good point, a negative (or too large) CIE offset in an `.eh_frame` entry header would cause this code:

` .eh_frame => stream.pos - id_len - id,`

to overflow. I'll put in a check for this.

Reply:
Do you want to put in a check in this PR or a follow-up? If you already did, I cannot seem to find it. It you didn't, I'm happy to merge the PR and you can add a check in a follow-up.

Reply:
I put that in my follow up commit here: https://github.com/ziglang/zig/pull/15823/commits/b1d86db7b45c57b2a9d48655738bce8d77327438#diff-3e42591e9375695ad5d2511e8300c6c30ccf11a80b9f1ed9bc50aba1cfb0b397R2528 `dwarf.zig:2528`

Reply:
Excellent, thanks!